src/libraries/System.Private.Uri/tests/FunctionalTests/PercentEncodingHelperTests.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System.Collections.Generic;
+using System.Tests;
+using Xunit;
+
+namespace System.PrivateUri.Tests
+{
+    public class PercentEncodingHelperTests
+    {
+        private const string OneByteUtf8 = "%41";           // A
+        private const string TwoByteUtf8 = "%C3%BC";        // ü
+        private const string ThreeByteUtf8 = "%E8%AF%B6";   // 诶
+        private const string FourByteUtf8 = "%F0%9F%98%80"; // 😀
+
+        private const string InvalidOneByteUtf8 = "%FF";
+        private const string OverlongTwoByteUtf8 = "%C1%81";        // A
+        private const string OverlongThreeByteUtf8 = "%E0%83%BC";   // ü
+        private const string OverlongFourByteUtf8 = "%F0%88%AF%B6"; // 诶;
+
+        public static IEnumerable<object[]> PercentEncodedAndDecodedUTF8Sequences()
+        {
+            static object[] Pair(string s1, string s2) => new object[] { s1, s2 };
+
+            yield return Pair(OneByteUtf8, "A");
+            yield return Pair(TwoByteUtf8, "ü");
+            yield return Pair(ThreeByteUtf8, "诶");
+            yield return Pair(FourByteUtf8, "😀");
+
+            yield return Pair(OneByteUtf8 + OneByteUtf8, "AA");
+            yield return Pair(TwoByteUtf8 + TwoByteUtf8, "üü");
+            yield return Pair(ThreeByteUtf8 + ThreeByteUtf8, "诶诶");
+            yield return Pair(FourByteUtf8 + FourByteUtf8, "😀😀");
+
+            yield return Pair(OneByteUtf8 + TwoByteUtf8 + OneByteUtf8, "AüA");
+            yield return Pair(TwoByteUtf8 + ThreeByteUtf8 + TwoByteUtf8, "ü诶ü");
+
+            yield return Pair(InvalidOneByteUtf8 + OneByteUtf8, InvalidOneByteUtf8 + "A");
+            yield return Pair(OverlongTwoByteUtf8 + TwoByteUtf8, OverlongTwoByteUtf8 + "ü");
+            yield return Pair(OverlongThreeByteUtf8 + ThreeByteUtf8, OverlongThreeByteUtf8 + "诶");
+            yield return Pair(OverlongFourByteUtf8 + FourByteUtf8, OverlongFourByteUtf8 + "😀");
+
+            yield return Pair(InvalidOneByteUtf8, InvalidOneByteUtf8);
+            yield return Pair(InvalidOneByteUtf8 + InvalidOneByteUtf8, InvalidOneByteUtf8 + InvalidOneByteUtf8);
+            yield return Pair(InvalidOneByteUtf8 + InvalidOneByteUtf8 + InvalidOneByteUtf8, InvalidOneByteUtf8 + InvalidOneByteUtf8 + InvalidOneByteUtf8);
+
+            // 11001010 11100100 10001000 10110010 - 2-byte marker followed by 3-byte sequence
+            yield return Pair("%CA" + "%E4%88%B2", "%CA" + '䈲');
+
+            // 4 valid UTF8 bytes followed by 5 invalid UTF8 bytes
+            yield return Pair("%F4%80%80%BA" + "%FD%80%80%BA%CD", char.ConvertFromUtf32(0x10003A) + "%FD%80%80%BA%CD");
```

Comment:
Nit: you can replace `char.ConvertFromUtf32(0x10003A)` with the literal `"\U0010003A"`. `"\U..."` (as opposed to `"\u..."`) is a convenient way to write supplementary plane values. :)

(Same comment later in this method.)

Reply:
Thanks :) Will change it

Reply:
I wasn't aware of the difference between \u \U before, thanks for pointing it out